File: app/src/main/java/com/rpifinal/hitema/controller/UpdateUserActivity.java
package com.rpifinal.hitema.controller;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.tasks.OnSuccessListener;
import com.rpifinal.hitema.R;
import com.rpifinal.hitema.model.User;
import api.UserHelper;
import butterknife.BindView;
import butterknife.OnClick;

//TODO : Optimisation REGEX

public class UpdateUserActivity extends BaseActivity {

    // =============================================================================================
    // ATTRIBUTS MEMBRES

    private static final String TAG = "UpdateUserActivity";

    // Expressions régulières
    private static final String REGEX_FL_NAME = "[a-zA-Z-]+[:blank]?[a-zA-Z]+";
    private static final String REGEX_USERNAME = "[a-zA-Z0-9_]+";

    // CODE REST
    private static final int UPDATE_USERNAME  = 10;
    private static final int UPDATE_FIRSTNAME = 20;
    private static final int UPDATE_LASTNAME  = 30;

    // Binding UI
    @BindView(R.id.update_activity_title_textView)   TextView mTitleUpdateTextView;
    @BindView(R.id.update_activity_username_field)   EditText mUsernameUpdateField;
    @BindView(R.id.update_activity_firstName_field)  EditText mFirstnameUpdateField;
    @BindView(R.id.update_activity_lastName_field)   EditText mLastnameUpdateField;
    @BindView(R.id.update_activity_username_submit)  Button mUsernameButton;
    @BindView(R.id.update_activity_firstName_submit) Button mFirstnameButton;
    @BindView(R.id.update_activity_lastName_submit)  Button mLastnameButton;
    // =============================================================================================

    // =============================================================================================
    @Override
    protected void onCreate(Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);
        this.updateUI();
    }

    @Override
    public int getFragmentLayout() {

        return R.layout.activity_update_user;
    }
    // =============================================================================================

    // =============================================================================================
    // ACTIONS

    // TODO : Regrouper onCLick

    //Boutton de retour de la page update vers la page profil
    @OnClick(R.id.back_to_profile)
    public void onClickReturnProfilButton() {

        Intent profile = new Intent(this, ProfileActivity.class);
        startActivity(profile);
    }

    @OnClick({R.id.update_activity_username_submit, R.id.update_activity_firstName_submit, R.id.update_activity_lastName_submit})
    public void onClickUpdate(View view) {
        String uid = getCurrentUser().getUid();
        switch(view.getId())
        {
            case R.id.update_activity_username_submit:
                String username = mUsernameUpdateField.getText().toString();
                checkDataEntry(UPDATE_USERNAME,username,uid);
                break;

            case R.id.update_activity_firstName_submit:
                String firstName = mFirstnameUpdateField.getText().toString();
                checkDataEntry(UPDATE_FIRSTNAME,firstName,uid);

                break;

            case R.id.update_activity_lastName_submit:
                String lastName = mLastnameUpdateField.getText().toString();
                checkDataEntry(UPDATE_LASTNAME,lastName,uid);
                break;
        }
        //Hide keyboard when click on the update button
        closeKeyboard();
    }

    //Method closeKeyboard() that will close/hide the keyboard when click on the valid update button
    private void closeKeyboard() {

        View view = this.getCurrentFocus();
        if (view != null)
        {
            InputMethodManager imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
            imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
        }
    }

    //Hide keyboard when click outside the screen
    /*
    public void hideKeyboard(View view) {
        InputMethodManager inn = (InputMethodManager) getSystemService(Activity.INPUT_METHOD_SERVICE);
        inn.hideSoftInputFromWindow(view.getWindowToken(),0);
    }
    */

    // =============================================================================================
    // UI/UX

    public void checkDataEntry(int code,String data,String uid) {

        final boolean namesCondition = data.matches(REGEX_FL_NAME) && !data.equals("") && data.length() > 3;

        switch (code)
        {
                // =================================================================================
                case UPDATE_LASTNAME:

                    if (namesCondition)
                    {
                        String successMessage = getString(R.string.success_update_lastname);
                        UserHelper.updateLastName(data, uid).addOnFailureListener(this.onFailureListener())
                                .addOnSuccessListener(this.onSuccessListener(successMessage));
                    }
                    else
                    {
                        String errorMessage = getString(R.string.error_update_lastname);
                        Toast.makeText(getApplicationContext(), errorMessage, Toast.LENGTH_SHORT).show();
                    }
                    break;
                // =================================================================================
                case UPDATE_FIRSTNAME:

                    if (namesCondition)
                    {
                        String successMessage = getString(R.string.success_update_firstname);
                        UserHelper.updateFirstName(data, uid).addOnFailureListener(this.onFailureListener())
                                .addOnSuccessListener(this.onSuccessListener(successMessage));
                    }
                    else
                    {
                        String errorMessage=getString(R.string.error_update_firstname);
                        Toast.makeText(getApplicationContext(), errorMessage, Toast.LENGTH_SHORT).show();
                    }
                    break;
                // =================================================================================
                case UPDATE_USERNAME:

                    if (data.matches(REGEX_USERNAME) && !data.equals("") && data.length()> 3)
                    {
                        String successMessage = getString(R.string.success_update_username);
                        UserHelper.updateUsername(data, uid).addOnFailureListener(this.onFailureListener()).
                                addOnSuccessListener(this.onSuccessListener(successMessage));
                    }
                    else
                    {
                        String errorMessage=getString(R.string.error_update_username);
                        Toast.makeText(getApplicationContext(), errorMessage, Toast.LENGTH_SHORT).show();
                    }
                    break;
                // =================================================================================
        }
            updateUI();
    }

    public void updateUI() {

        if(this.getCurrentUser() != null)
        {
            UserHelper.getUser(this.getCurrentUser().getUid()).addOnSuccessListener(documentSnapshot -> {

                User currentUser = documentSnapshot.toObject(User.class);

                if(!currentUser.getUsername().isEmpty())
                {
                    this.mUsernameUpdateField.setText(currentUser.getUsername());
                }

                if(!currentUser.getFirstName().isEmpty())
                {
                    this.mFirstnameUpdateField.setText(currentUser.getFirstName());
                }

                if(!currentUser.getLastName().isEmpty())
                {
                    this.mLastnameUpdateField.setText(currentUser.getLastName());
                }
            });
        }

        if(this.getCurrentUser() == null)
        {
            this.mTitleUpdateTextView.setText(getString(R.string.no_user_connected_error));
        }
    }

    public void makeToast() {

        Toast.makeText(getBaseContext(), "OKAY", Toast.LENGTH_SHORT).show();
    }

    private OnSuccessListener onSuccessListener(String resId) {

        return e-> Toast.makeText(getApplicationContext(), resId, Toast.LENGTH_SHORT).show();
    }
    // =============================================================================================
}

File: app/src/main/java/oldFiles/UpdateUserActivity.kt
package oldFiles

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import oldFiles.api.UserHelper.getUser
import oldFiles.api.UserHelper.updateFirstName
import oldFiles.api.UserHelper.updateLastName
import oldFiles.api.UserHelper.updateUsername
import com.google.android.gms.tasks.OnSuccessListener
import com.google.firebase.firestore.DocumentSnapshot
import com.rpifinal.hitema.R
import oldFiles.model.User

//TODO : Optimisation REGEX
class UpdateUserActivity : BaseActivity() {
    // Binding UI
    @JvmField
    //@BindView(R.id.update_activity_title_textView)
    var mTitleUpdateTextView: TextView? = null
    @JvmField
    //@BindView(R.id.update_activity_username_field)
    var mUsernameUpdateField: EditText? = null
    @JvmField
    //@BindView(R.id.update_activity_firstName_field)
    var mFirstnameUpdateField: EditText? = null
    @JvmField
    //@BindView(R.id.update_activity_lastName_field)
    var mLastnameUpdateField: EditText? = null
    @JvmField
    //@BindView(R.id.update_activity_username_submit)
    var mUsernameButton: Button? = null
    @JvmField
    //@BindView(R.id.update_activity_firstName_submit)
    var mFirstnameButton: Button? = null
    @JvmField
    //@BindView(R.id.update_activity_lastName_submit)
    var mLastnameButton: Button? = null

    // =============================================================================================
    // =============================================================================================
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        updateUI()
    }

    override val fragmentLayout: Int
        get() = R.layout.activity_update_user

    // =============================================================================================
    // =============================================================================================
    // ACTIONS
    // TODO : Regrouper onCLick
    //Boutton de retour de la page update vers la page profil
    //@OnClick(R.id.back_to_profile)
    fun onClickReturnProfilButton() {
        val profile = Intent(this, ProfileActivity::class.java)
        startActivity(profile)
    }

    //@OnClick(R.id.update_activity_username_submit, R.id.update_activity_firstName_submit, R.id.update_activity_lastName_submit)
    fun onClickUpdate(view: View) {
        val uid = currentUser!!.uid
        when (view.id) {
            R.id.update_activity_username_submit -> {
                val username = mUsernameUpdateField!!.text.toString()
                checkDataEntry(UPDATE_USERNAME, username, uid)
            }
            R.id.update_activity_firstName_submit -> {
                val firstName = mFirstnameUpdateField!!.text.toString()
                checkDataEntry(UPDATE_FIRSTNAME, firstName, uid)
            }
            R.id.update_activity_lastName_submit -> {
                val lastName = mLastnameUpdateField!!.text.toString()
                checkDataEntry(UPDATE_LASTNAME, lastName, uid)
            }
        }
        //Hide keyboard when click on the update button
        closeKeyboard()
    }

    //Method closeKeyboard() that will close/hide the keyboard when click on the valid update button
    private fun closeKeyboard() {
        val view = this.currentFocus
        if (view != null) {
            val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            imm.hideSoftInputFromWindow(view.windowToken, 0)
        }
    }

    //Hide keyboard when click outside the screen
/*
    public void hideKeyboard(View view) {
        InputMethodManager inn = (InputMethodManager) getSystemService(Activity.INPUT_METHOD_SERVICE);
        inn.hideSoftInputFromWindow(view.getWindowToken(),0);
    }
    */
    // =============================================================================================
    // UI/UX
    fun checkDataEntry(code: Int, data: String, uid: String?) {
        val namesCondition = data.matches(REGEX_FL_NAME.toRegex()) && data != "" && data.length > 3
        when (code) {
            UPDATE_LASTNAME -> if (namesCondition) {
                val successMessage = getString(R.string.success_update_lastname)
                updateLastName(data, uid).addOnFailureListener(onFailureListener())
                        .addOnSuccessListener(onSuccessListener(successMessage) as OnSuccessListener<in Void>)
            } else {
                val errorMessage = getString(R.string.error_update_lastname)
                Toast.makeText(applicationContext, errorMessage, Toast.LENGTH_SHORT).show()
            }
            UPDATE_FIRSTNAME -> if (namesCondition) {
                val successMessage = getString(R.string.success_update_firstname)
                updateFirstName(data, uid).addOnFailureListener(onFailureListener())
                        .addOnSuccessListener(onSuccessListener(successMessage) as OnSuccessListener<in Void>)
            } else {
                val errorMessage = getString(R.string.error_update_firstname)
                Toast.makeText(applicationContext, errorMessage, Toast.LENGTH_SHORT).show()
            }
            UPDATE_USERNAME -> if (data.matches(REGEX_USERNAME.toRegex()) && data != "" && data.length > 3) {
                val successMessage = getString(R.string.success_update_username)
                updateUsername(data, uid).addOnFailureListener(onFailureListener()).addOnSuccessListener(onSuccessListener(successMessage) as OnSuccessListener<in Void>)
            } else {
                val errorMessage = getString(R.string.error_update_username)
                Toast.makeText(applicationContext, errorMessage, Toast.LENGTH_SHORT).show()
            }
        }
        updateUI()
    }

    fun updateUI() {
        if (this.currentUser != null) {
            getUser(currentUser!!.uid).addOnSuccessListener { documentSnapshot: DocumentSnapshot ->
                val currentUser = documentSnapshot.toObject(User::class.java)
                if (!currentUser!!.username?.isEmpty()!!) {
                    mUsernameUpdateField?.setText(currentUser.username)
                }
                if (!currentUser.firstName?.isEmpty()!!) {
                    mFirstnameUpdateField?.setText(currentUser.firstName)
                }
                if (!currentUser.lastName?.isEmpty()!!) {
                    mLastnameUpdateField?.setText(currentUser.lastName)
                }
            }
        }
        if (this.currentUser == null) {
            mTitleUpdateTextView!!.text = getString(R.string.no_user_connected_error)
        }
    }

    fun makeToast() {
        Toast.makeText(baseContext, "OKAY", Toast.LENGTH_SHORT).show()
    }

    private fun onSuccessListener(resId: String): OnSuccessListener<*> {
        return OnSuccessListener { e: Any? -> Toast.makeText(applicationContext, resId, Toast.LENGTH_SHORT).show() }
    } // =============================================================================================

    companion object {
        // =============================================================================================
        // ATTRIBUTS MEMBRES
        private const val TAG = "UpdateUserActivity"
        // Expressions régulières
        private const val REGEX_FL_NAME = "[a-zA-Z-]+[:blank]?[a-zA-Z]+"
        private const val REGEX_USERNAME = "[a-zA-Z0-9_]+"
        // CODE REST
        private const val UPDATE_USERNAME = 10
        private const val UPDATE_FIRSTNAME = 20
        private const val UPDATE_LASTNAME = 30
    }
}